### ISSUE TITLE
Update documentation and make it markdown

### ISSUE THREAD

**Issue by ssardina** [SWI-Prolog/packages-jpl#22]
Currently doc in the repo is quite old (JPL <= 3) and does not reflect the current JPL V7.

@anionic has however a nice HTML doc in https://jpl7.org/ 

He has already done major updates to sync with the latest version 

As per discussion in PR #5, we want to bring that doc into Markdown language

**Comment by ssardina** [comment]
Yes that was my original idea. To replace everything that is inside `doc/` in the main repo that you have. Even more I think the best is to do this as a Wiki which itself is a git repo and we can add it inside `doc/` 

However, @anionic may have suggested other plans, something around Github Projects but I quite don't follow because Github Projects are not for this. Wikies are.

**Comment by pmoura** [comment]
His suggestion is to use **GitHub Pages** . See https://pages.github.com It's a good choice for this case.

**Comment by ssardina** [comment]
ahhh that makes total sense now... :-)

**Comment by ssardina** [comment]
OK I have made substantial progress on the github page, including migrating quite a bit of jpl7.org and recovering stuff from the very old doc:

https://ssardina-research.github.io/packages-jpl/

Let me know if this is what you had in mind @anionic . Of course quite a bit more work is needed, but I want to make sure others are happy with this direction before continuing...

(Would be nice to include as much doc as possible for people to get it easy to start-up. Also do we have some list of projects that use JPL? I have mine, but do we know of some?)

**Comment by JanWielemaker** [comment]
I guess what matters is that there is no (not much) outdated stuff around, the new stuff is in a place where people easily find it and where the source is safe.   If the source is now markdown we could also merge it into the package description which now only covers the Prolog side of the API.  I guess the JavaDoc documentation complicates this.  Surely we should make sure there is a link that is not easily missed.

**Comment by anionic** [comment]
That's great! I'm looking forward to writing Markdown not HTML ;-)
Where is the generated TOC? Can we have a hierarchy of page links on the RHS so users can go from any page directly to any other?
Some of the newer pages are very long, with links at the top to anchors within, I guess this is poor style and they should be split into separate pages...
If you want to try out the custom domain before we go live with jpl7.org, use anionic.co.uk and I'll do the DNS thing at my end.

**Comment by ssardina** [comment]
Guys, GitHub pages are cool but if you want to do something more than the default themes it requires quite a good understanding of [jekyll](https://github.com/jekyll/jekyll).

There are lots of themes one can pick outside the built-in in the github interfaces, and some seem to have sidebars, but they are not easy to understand or configure. For example, one very neat is [minimal-mistakes](https://mmistakes.github.io/minimal-mistakes/)

Adding a sidebar to the one I have does not appear easy...

**Comment by anionic** [comment]
re: domain: I suggested my spare anionic.co.uk so we could get this working without breaking the current live jpl7.org site, but anyway I've replaced jpl7.org's old DNS A record with the four GitHub Pages IPv4 addresses, and these seem to have propagated somewhat. GitHub docs say you'll need to "remove and then re-add your custom domain to the repository you’re using to publish your Pages site to trigger the process of enabling HTTPS".
re: styling and navigation: can we keep it simple, avoiding anyone else's code? Aren't we just styling some headings, prose, links, code fragments and the odd table? Sections 7-9 of http://jmcglone.com/guides/github-pages/ shows how to use a common page template for uniform navigation; I can contribute a simple responsive TOC in a RH sidebar above some viewport width, or dropdown below that width. We can maintain the TOC manually.

**Comment by ssardina** [comment]
Yeap, I would like to keep it as simple as possible and avoid huge infrastructures from others for this. I first though the easier would be to just use the built-in themes from Github, but these do not seem to have sidebar.

At this point I have the pack of pages in the Github Wiki:

https://github.com/ssardina-research/packages-jpl/wiki

I will have a look at the link you sent, thanks. But wouldn't be enough if we keep a Wiki in the main JPL repo (the wiki itself is also a git repo by itself)? 

No idea what is happening with the DNS, but now it is working: http://anionic.co.uk/

Actually, now that I look it it, it looks neat and simple. Nicer than the wiki itsef. if we could just have a sidebar there...

**Comment by JanWielemaker** [comment]
> At this point I have the pack of pages in the Github Wiki:

To me, this looks good enough.  It seems pretty clean and it contains far fewer errors, which is the most important part.  If I'd want anything, I'd like to have it running together with the rest of the docs on http://swi-prolog.org.   For now, a clear intro in the package description that tells you what to find at the wiki is just fine, I'd say.

**Comment by ssardina** [comment]
Yes Jan, agree. I had too many doc versions floating around (wiki, standalone html, github pages, etc) and it was becoming confusing.

Since I do not have much more time and following your suggestion what I did was to clean-up the doc I have so far in the Wiki of my fork:

https://github.com/ssardina-research/packages-jpl/wiki

This has the best version I could produce so far, with lots of old stuff either dropped or updated, and some new information, mostly to help anyway new to JPL.

I spent some time looking at GitHub pages and I stopped as I won't have the time to do it now and wanted to spend the time in cleaning up the doc. What I can say is that:

1. Github pages with Jekyll is fantastic, clean and powerful.
2. The default "out of the box" themes are not very versatile, not much advantage than the wiki except colors
3. If you want to do a bit more, like having _sidebars then it does require some work. 
    * Doing a whole template site from scratch requires time. Not hard but lots of time to get the first version.
    * There are amazing clean Jekyll themes the one can use. After looking around I would use this:
https://idratherbewriting.com/documentation-theme-jekyll/

However, at this stage, do we need a funky web page? A good quality wiki will do it maybe, and we may as well spend time getting JPL more powerful.

Happy to hear your views...

**Comment by JanWielemaker** [comment]
I think we should keep it at this.   If you want to spend more cycles I guess improving JPL itself is a better use of your time.   I guess this means I can delete all `.html` files from the jpl git and add a simple readme pointing at the places where the docs are?   Or, as you have it all in your head, could you provide a quick pull request for that?

**Comment by ssardina** [comment]
Agree & done! 

Not sure if you meant PR the wiki to your site, don't know how to do it. As for now, while I am still developing and polishing it, I put a minimal README pointing to my Wiki. All old .html files are gone. 

left `docs/` with a simple `README.md` inside because we may use that subdir in the (near future).

Is that more or less what you wanted? I think it is now decent to go live, I feel comfortable with the Wiki even though we can still polish it better (need to migrate the release notes of jpl7.org for example)

@anionic , I wanted ot still make a reference to `jpl7.org` but it is giving me this:

`You cannot visit jpl7.org right now because the website uses HSTS. Network errors and attacks are usually temporary, so this page will probably work later.

`

**Comment by pmoura** [comment]
GitHUb wikis are git repos and thus can be cloned. Thus, one solution would be for Jan to clone the wiki and then copy the files and git add them elsewhere. Thanks for your contributions to improve JPL!

**Comment by ssardina** [comment]
1. Indeed, they are just repos and that's how I work locally and then push it.
2. That's also why I left `docs/` to probably store the wiki there? But since I have not yet finished with it I thought we may as well do a first clean-up and a short `README.md`
3. Jan and others, I forgot, please check the LICENSE that I added, it is the BSD 2-Clause "Simplified" License as per original JPL, but just in case check it.
4. @pmoura , my pleasure, thanks to you for doing such a neat framework! 
5. Do we know any "projects" that have used or use JPL? Mine is within multi-agent systems ([here](https://bitbucket.org/ssardina-research/sarl-prolog-cap)) but would be nice to list a few if we know them...

**Comment by pmoura** [comment]
@ssardina I did contract work for companies where I used Logtalk + SWI-Prolog + JPL solutions. Usually to interface with popular Java libraries (e.g. Apache Commons). But so far no public/open source projects  where I used these solutions.

**Comment by ssardina** [comment]
Good, I like it. Maybe to the two links you want to add another one to https://jpl7.org/ as "Java Doc & Release notes". (those two have not yet been incorporated into the Wiki, will do it soon). 

> Does it makes sense to migrate the wiki to the central packages-jpl wiki?

YES, I have no problem with that Jan. I am just " working"  on it and once it is there I will not be able to modify it?  mmm how can we do it so we can incrementally work on the doc? The Wiki is just another repo, but you cannot give particular user access to it independently I think... One way is to add the Wiki repo itself inside packages-jpl `doc/` subdir? I don' t know..

**Comment by ssardina** [comment]
> Are you now proposing to publish the docs as a publicly editable wiki? 

Jan seems to suggest we put the Wiki in the `packages-jpl`  repo, which makes total sense, everything in one place.

What I would like, at least now, is that people like you or me are able to edit it to improve and extend it. Not sure how we can do that in the Wiki of Github 

> at jpl7.org or where? 

My understanding is that everything in once place: `packages-jpl` github repo.

> are the APi pages from the old jpl7.org included?

Yes, and they include the last big improvements you did to bring the doc in sync to the latest version:

https://github.com/ssardina-research/packages-jpl/wiki/Java-API
https://github.com/ssardina-research/packages-jpl/wiki/Prolog-API

Basically I converted the `html` to markdown using the tool and then manual editing.

The Wiki is to me in a "reasonable" publishable state but I think we want to:

1. include the release notes
2. accommodate the javadoc (we also need to update them)
3. keep refining the API docs, there are still glitches


That' s why I would like to still have easy access to edit....

Is there any particular way you propose to go ahead @anionic? I agree with Jan that it is best to spend the energy improving what we have and in particular the JPL code itself, as we have done in the past weeks.

**Comment by JanWielemaker** [comment]
> Jan seems to suggest we put the Wiki in the `packages-jpl` repo, which makes total sense, everything in one place.

I was suggesting to migrate the wiki on your clone of packages-jpl to the main one.  Yes, by default wikis are publicly writeable (with a github account).  I've rarely seen them being misused.  There is also a flag that allows writing to them only for users with write access to the repo itself.  I propose to add @ssardina to access list of the main packages-jpl.git.  Ok?  I'm happy to take care of migrating the wiki.

**Comment by anionic** [comment]
Surely, all documentation should be maintained within packages-jpl, where the wiki is for _developer_ documentation; _user_ documentation should continue to be published at jpl7.org (it's just a domain name) using GitHub Pages, with a custom URL, from the docs subfolder. I've looked into this, and GitHub wiki content cannot (easily) be published via Pages. I am happy for the developer wiki to be writeable by any GitHub user; only packages-jpl contributors (gratefully including our newest recruit @ssardina!) will be able to alter the user docs, but that's ok. Until we have got this working, I propose using the spare domain name anionic.co.uk as the custom URL (@ssardina will need to remove this as a custom URL for his clone); I'll continue to publish my "JPL 7.4.0" docs at jpl7.org from our own server, then switch over when we're ready. Everything will then be GitHub-based and idiomatic.

**Comment by ssardina** [comment]
> Surely, all documentation should be maintained within packages-jpl, where the wiki is for _developer_ documentation; _user_ documentation should continue to be published at jpl7.org (it's just a domain name) using GitHub Pages, with a custom URL, from the docs subfolder. I've looked into this, and 

Good. The Wiki now has both _developer_ and _user_ doc. But it can be easily "refactored"

> GitHub wiki content cannot (easily) be published via Pages. I am happy for the developer wiki to be writeable by any GitHub user; only packages-jpl contributors (gratefully including our newest recruit @ssardina!) will be able to alter the user docs, but that's ok. 

Didn't know that Wiki repos can be changed by any GitHub user. Strange for Doc but...

> Until we have got this working, I propose using the spare domain name anionic.co.uk as the custom URL (@ssardina will need to remove this as a custom URL for his clone); I'll continue to publish my "JPL 

I have just removed it.

> 7.4.0" docs at jpl7.org from our own server, then switch over when we're ready. Everything will then be GitHub-based and idiomatic.

Sounds like a plan. So, to recap, at this point we have:

* Decent all together doc user+developer in MY Wiki, and a simple README and pointer in the main `packages-jpl` site. Also a clean-up `jpl7.org` version aligned with the Wiki for the APIs.

... and we would like in the end to:

1. Move the developer part to the Wiki in official Github `packages-jpl` site. Any github user can edit. 
2. Create a Github pages under `doc/` subdir in official `packages-jpl` that we can edit to host the _user_ documentation. Only `packages-jpl` contributors will be able to edit this. This will have the domain name `jpl7.org`.

Now, the only issue is that, while not difficult, step 2 will take some effort and time unless you are already familiar with Github Pages and themes.  Following @anionic link, I tested and created a blank template
`https://ssardina.github.io/` to play around. It is not hard at all, but to get to the page we want will take time for me. For example, managing css styles and adding a sidebar, they are possible, but I just don't know it and would have to learn it (not that I enjoy building web pages!). Personally, I prefer to spend my few CPU cycles investing on JPL features, for example some of the issues I created, given that we now have a decent clean-up pack of doc.

@anionic : do you have enough knowledge of Github Pages theme to create a framework where we can add pages with Markdown? As you said, we do not want to write html pages. If I have the framework, with the sidebar, etc, I can move the user info of the Wiki there and even add incrementally things that I think will be useful for the user (e.g., specific examples how you do things).

**Comment by anionic** [comment]
Yes I can do step 2. I have tried this out in a dummy repository using another spare domain name. As it only involves text files, I find it convenient to set this up using the web interface to GitHub, committing directly to master. @JanWielemaker are you comfortable with me doing this to SWI-Prolog / packages-jpl?!
@ssardina do you have any tips about using pandoc to convert my HTML docs?

**Comment by ssardina** [comment]
That is great @anionic so we can split the work well :-)

You can set the web page to be `jpl.github.io` for example.

However, if you do this under the `doc` subidr, you need to do it in the official `packages-jpl` repo, as you know. So @JanWielemaker has to be happy with that. And it has to be against master branch, seems you cannot "preview" the Github pages while it is on a branch! :-(

Regarding pandoc, there is not much. I just ran this:

` pandoc -f html -t markdown JPL\ -\ Release\ notes\ -\ 7.4.0.html`

It produces a very good first take, but then I did manual editing, quite a bit but doable. @anionic I can do that at least with many pages. What I would benefit is that you set-up the "framework templte" of the web page, with the sidebar and placeholder files. Then I can insert many pages, many have been curated already, I can do that.

What do you think?

Actually, that would require me to be able to write into jpl repo.. an alternative would be we create a new repo called 'doc-jpl' ? just for the web-page?

**Comment by JanWielemaker** [comment]
I've added @ssardina to the jpl team, so you both have full access to the module.  You both know much more about JPL than I do, so please do as you seem fit.  I only care where it concerns building, testing and distributing the package.    Some people with write access create a branch and pull request, others update master.  For documentation I surely recommend to update master directly.  If you change code, please make sure it still builds and runs the tests.

Building will move to `cmake`.  At the moment I still maintain both build systems.   Not sure how soon that will change.   Looking forward to a nicer JPL!

**Comment by ssardina** [comment]
Excellent. Thanks Jan. 

Will keep the conversation on the JPL member's chat

**Comment by ssardina** [comment]
We can close this, we now have the doc in a github page and under `doc/`.
